File: catch.py
import abc
import collections
import functools
import random

import matplotlib.pyplot as plt


class Environment(object):
  __metaclass__ = abc.ABCMeta

  @abc.abstractmethod
  def available_actions(self):
    return []

  @abc.abstractmethod
  def reset(self):
    pass

  @abc.abstractmethod
  def observe(self):
    return []

  @abc.abstractmethod
  def act(self, action):
    """Returns (reward, is_terminal)."""
    return 0.0, True


class Catch(Environment):

  def __init__(self, width):
    self.width = width
    self.height = width
    self.reset()

  def available_actions(self):
    return [-1, 0, 1]

  def reset(self):
    self.ball_position = [0, random.randint(0, self.width - 1)]
    self.bat_position = random.randint(0, self.width - 1)

  def observe(self):
    return self._render(self.ball_position[0], self.ball_position[1],
                        self.bat_position)

  def _render(self, ball_h, ball_w, bat_w):
    screen = []
    for h in range(self.height):
      row = []
      for w in range(self.width):
        if h == ball_h and w == ball_w:
          row.append('*')
        elif w == self.bat_position and h == self.height - 1:
          row.append('_')
        else:
          row.append(' ')
      screen.append(row)
    return screen

  def act(self, action):
    self.bat_position = max(0, min(self.width - 1, self.bat_position + action))
    self.ball_position[0] += 1
    if self.ball_position[0] == self.height - 1:
      if self.bat_position == self.ball_position[1]:
        return 1, True
      else:
        return -1, True
    return 0, False


class QLearner(object):
  """Off-Policy TD Control, page 158."""

  def __init__(self, available_actions, learning_rate, discount_factor):
    # action_value_estimates
    self.Q = collections.defaultdict(lambda: collections.defaultdict(int))
    self.available_actions = available_actions
    self.learning_rate = learning_rate
    self.discount_factor = discount_factor

  def act(self, observation):
    q_action = self.Q[self._condense_observation(observation)]
    return functools.reduce(lambda a, b: a if q_action[a] > q_action[b] else b,
                            self.available_actions)

  def learn(self, observation_raw, action, new_observation_raw, reward):
    observation = self._condense_observation(observation_raw)
    new_observation = self._condense_observation(new_observation_raw)
    max_q = max([self.Q[new_observation][a] for a in self.available_actions])
    self.Q[observation][action] += self.learning_rate * \
        (reward + self.discount_factor * max_q - self.Q[observation][action])

  def _condense_observation(self, observation):
    return '\n'.join([''.join(row) for row in observation])

  def __str__(self):
    return 'GreedyAgent'

num_episodes = 350
num_repetitions = 200

plt.ion()
x = range(num_episodes)
rewards = [0 for _ in range(num_episodes)]
fig = plt.figure()
ax = fig.add_subplot(111)
ax.set_ylim([-1.1, 1.1])
line = ax.plot(x, rewards, '-')[0]


for repetition in range(num_repetitions):
  catch = Catch(5)
  agent = QLearner(catch.available_actions(), learning_rate=0.1,
                   discount_factor=0.9)
  reward_ma = 0
  for i in range(num_episodes):
    terminal = False
    catch.reset()

    while not terminal:
      observation = catch.observe()
      action = agent.act(observation)
      reward, terminal = catch.act(action)
      new_observation = catch.observe()
      agent.learn(observation, action, new_observation, reward)

    reward_ma = 0.05 * reward + 0.95 * reward_ma

    rewards[i] += 1.0 / (repetition + 1) * (reward_ma - rewards[i])
  if repetition % 3 == 0:
    line.set_ydata(rewards)
    fig.canvas.draw()

raw_input()

File: catch.lua
require 'nn'
local gfx = require 'gfx.js'

function defaultdict(default_value_factory)
  local t = {}
  local metatable = {}
  metatable.__index = function(t, key)
    if not rawget(t, key) then
      rawset(t, key, default_value_factory(key))
    end
    return rawget(t, key)
  end
  return setmetatable(t, metatable)
end

function reduce(f, xs)
  acc = xs[1]
  for i = 2, #xs do
    acc = f(acc, xs[i])
  end
  return acc
end

local Catch = {}
Catch.__index = Catch

function Catch.create(width)
  local catch = {}
  setmetatable(catch, Catch)
  catch.width = width
  catch.height = width
  return catch
end

-- actions
function Catch:availableActions()
  return {-1, 0, 1}
end

function Catch:reset()
  self.ballPosition = {0, math.floor(torch.uniform(0, self.width))}
  self.batPosition = math.floor(torch.uniform(0, self.width))
end

-- observations
function Catch:observe()
  return torch.Tensor({self.ballPosition[1], self.ballPosition[2], self.batPosition})
end

-- reward, isTerminal
function Catch:act(action)
  self.batPosition = math.max(0, math.min(self.width - 1, self.batPosition + action))
  self.ballPosition[1] = self.ballPosition[1] + 1
  if self.ballPosition[1] == self.height - 1 then
    if self.batPosition == self.ballPosition[2] then
      return 1.0, true
    else
      return -1.0, true
    end
  end
  return 0.0, false
end

local QLearner = {}
QLearner.__index = QLearner

function QLearner.create(availableActions, learningRate, discountFactor)
  local learner = {}
  setmetatable(learner, QLearner)
  learner.Q = defaultdict(function(_) return defaultdict(function(_) return 0 end) end)
  learner.availableActions = availableActions
  learner.learningRate = learningRate
  learner.discountFactor = discountFactor
  return learner
end

function observationToString(observation)
  return observation:__tostring()
end

function QLearner:act(observation)
  q_action = self.Q[observationToString(observation)]
  return reduce(function(a, b) return q_action[a] > q_action[b] and a or b end,
                self.availableActions)
end

function QLearner:learn(observationRaw, action, newObservationRaw, reward)
  local max_q = - math.huge
  for i, a in pairs(self.availableActions) do
    local q = self.Q[observationToString(newObservationRaw)][a]
    if q > max_q then
      max_q = q
    end
  end
  local observation = observationToString(observationRaw)
  self.Q[observation][action] = (self.Q[observation][action] + self.learningRate *
    (reward + self.discountFactor * max_q - self.Q[observation][action]))
end

local NNLearner = {}
NNLearner.__index = NNLearner

function NNLearner.create(availableActions, learningRate, discountFactor)
  local learner = {}
  setmetatable(learner, NNLearner)
  learner.availableActions = availableActions
  learner.learningRate = learningRate
  learner.discountFactor = discountFactor

  learner.mlp = nn.Sequential();  -- make a multi-layer perceptron
  local inputs = 3 + #availableActions; local outputs = 1; local HUs = 20; -- parameters
  learner.mlp:add(nn.Linear(inputs, HUs))
  learner.mlp:add(nn.Tanh())
  learner.mlp:add(nn.Linear(HUs, outputs))
  return learner
end

function onehot(n, i)
  local xs = torch.zeros(n)
  xs[i] = 1
  return xs
end

function NNLearner:q(observation, action_index)
  return self.mlp:forward(torch.cat(observation,
                                    onehot(#self.availableActions, action_index)))
end

function NNLearner:act(observation)
  local max_q = - math.huge
  local action_index = -1
  for i, a in pairs(self.availableActions) do
    local q = self:q(observation, i)[1]
    if q > max_q then
      max_q = q
      action_index = i
    end
  end
  return self.availableActions[action_index]
end

function NNLearner:learn(observation, action, newObservation, reward)
  local max_q = - math.huge
  local action_index = -1
  for i, a in pairs(self.availableActions) do
    local q = self:q(newObservation, i)[1]
    if q > max_q then
      max_q = q
    end
    if a == action then
      action_index = i
    end
  end

  local target = reward + self.discountFactor * max_q - self:q(observation, action_index)[1]
  local criterion = nn.MSECriterion()

  local input = torch.cat(observation, onehot(#self.availableActions, action_index))
  -- feed it to the neural network and the criterion
  criterion:forward(self:q(observation, action_index),
                    torch.Tensor({target}))

  -- train over this example in 3 steps
  -- (1) zero the accumulation of the gradients
  self.mlp:zeroGradParameters()
  -- (2) accumulate gradients
  self.mlp:backward(input,
                    criterion:backward(self.mlp.output, torch.Tensor({target})))
  -- (3) update parameters with a 0.01 learning rate
  self.mlp:updateParameters(self.learningRate)
end

function formatData(rewards)
  return {
    {
      key = 'QLearner',
      values = rewards[1],
    },
  }
end


local numRepetitions = 200
local numEpisodes = 350
local rewards = torch.zeros(numEpisodes)
gfx.clear()
chartWindow = gfx.chart(rewards, {chart='line'})

for repetition = 1, numRepetitions do
  local catch = Catch.create(5)
  local agents = {
    -- QLearner.create(catch:availableActions(), 0.1, 0.9),
    NNLearner.create(catch:availableActions(), 0.01, 0.9),
  }

  for agent_i = 1, #agents do
    local agent = agents[agent_i]
    local reward_ma = 0
    for i = 1, numEpisodes do
      catch:reset()
      reward = 0.0
      terminal = false
      while not terminal do
        local observation = catch:observe()
        local action = agent:act(observation)
        reward, terminal = catch:act(action)
        local new_observation = catch:observe()
        agent:learn(observation, action, new_observation, reward)
        if terminal then
          break
        end
      end

      reward_ma = 0.05 * reward + 0.95 * reward_ma
      rewards[i] = (rewards[i] + 1.0 / (repetition + 1) *
        (reward_ma - rewards[i]))
    end
  end

  if repetition % 100 == 0 then
    gfx.chart(rewards, {chart='line', win=chartWindow})
  end
end
